core/node/proof_data_handler/src/tee_request_processor.rs

```diff
@@ -41,32 +38,71 @@ impl TeeRequestProcessor {
     ) -> Result<Json<TeeProofGenerationDataResponse>, RequestProcessorError> {
         tracing::info!("Received request for proof generation data: {:?}", request);
 
-        let mut connection = self
-            .pool
+        let mut min_batch_number: Option<L1BatchNumber> = None;
+
+        loop {
+            let l1_batch_number = match self
+                .lock_batch_for_proving(request.tee_type, min_batch_number)
+                .await?
+            {
+                Some(number) => number,
+                None => return Ok(Json(TeeProofGenerationDataResponse(None))),
+            };
+
+            match self.blob_store.get(l1_batch_number).await {
+                Ok(input) => {
+                    return Ok(Json(TeeProofGenerationDataResponse(Some(Box::new(input)))));
+                }
+                Err(ObjectStoreError::KeyNotFound(_)) => {
+                    tracing::warn!(
+                        "Blob for batch number {} has not been found in the object store.",
+                        l1_batch_number
+                    );
+                    self.unlock_batch(l1_batch_number, request.tee_type).await?;
+                    min_batch_number =
+                        min_batch_number.map_or(Some(l1_batch_number + 1), |num| Some(num + 1));
+                    continue;
+                }
+                Err(err) => {
+                    self.unlock_batch(l1_batch_number, request.tee_type).await?;
+                    return Err(RequestProcessorError::ObjectStore(err));
+                }
+            }
+        }
+    }
+
+    async fn lock_batch_for_proving(
+        &self,
+        tee_type: TeeType,
+        min_batch_number: Option<L1BatchNumber>,
+    ) -> Result<Option<L1BatchNumber>, RequestProcessorError> {
+        self.pool
             .connection()
             .await
-            .map_err(RequestProcessorError::Dal)?;
-
-        let l1_batch_number_result = connection
+            .map_err(RequestProcessorError::Dal)?
             .tee_proof_generation_dal()
-            .get_next_batch_to_be_proven(request.tee_type, self.config.proof_generation_timeout())
+            .lock_batch_for_proving(
+                tee_type,
+                self.config.proof_generation_timeout(),
+                min_batch_number,
+            )
             .await
-            .map_err(RequestProcessorError::Dal)?;
-
-        let l1_batch_number = match l1_batch_number_result {
-            Some(number) => number,
-            None => return Ok(Json(TeeProofGenerationDataResponse(None))),
-        };
+            .map_err(RequestProcessorError::Dal)
```

Comment:
Nit / off-topic (?): You can implement `From<DalError> for RequestProcessorError`, which would allow to skip such implicit conversions; `?` converts the error type automatically if possible.

Reply:
Good point, [fixed](https://github.com/matter-labs/zksync-era/pull/2674/commits/ef3395481a17618bd995e452b547f7faa7d9b6ab). ✅